src/application/service/AddResultsToTest.ts

```diff
@@ -0,0 +1,67 @@
+import { UserId } from '../../domain/model/user/UserId';
+import { Test } from '../../domain/model/test/Test';
+import { TestId } from '../../domain/model/test/TestId';
+import { TestTypeId } from '../../domain/model/testType/TestTypeId';
+import { TestRepository } from '../../domain/model/test/TestRepository';
+import { TestTypeRepository } from '../../domain/model/testType/TestTypeRepository';
+import { TestResultsCommand } from '../../api/interface';
+import { testResultsFactory } from './index';
+import { ResourceNotFoundError } from '../../domain/model/ResourceNotFoundError';
+import { TestType } from '../../domain/model/testType/TestType';
+import { User } from '../../domain/model/user/User';
+import { AccessDeniedError } from '../../domain/model/AccessDeniedError';
+import { DomainValidationError } from '../../domain/model/DomainValidationError';
+import { TestNotFoundError } from '../../domain/model/test/TestNotFoundError';
+import { AccessManagerFactory } from '../../domain/model/authentication/AccessManager';
+
+export class AddResultsToTest {
+  constructor(
+    private testRepository: TestRepository,
+    private testTypeRepository: TestTypeRepository,
+    private accessManagerFactory: AccessManagerFactory
+  ) {}
+
+  public async execute(actor: User, testId: string, resultsCommand: TestResultsCommand): Promise<Test> {
+    const test = await this.getTestOrFail(testId);
+    const testType = await this.getTestTypeOrFail(test.testTypeId);
+
+    await this.validateAccessAndPermissionToAddResults(actor, test, testType);
+
+    test.results = this.getResults(actor.id, testType, resultsCommand);
+    return this.testRepository.save(test);
+  }
+
+  private async getTestOrFail(testId: string) {
+    const test = await this.testRepository.findById(new TestId(testId));
+    if (!test) {
+      throw new TestNotFoundError(testId);
+    }
+    return test;
+  }
+
+  private async getTestTypeOrFail(testTypeId: TestTypeId) {
+    const testType = await this.testTypeRepository.findById(testTypeId);
+    if (!testType) {
+      throw new ResourceNotFoundError('testType', testTypeId.value);
+    }
+    return testType;
+  }
+
+  private async validateAccessAndPermissionToAddResults(actor: User, test: Test, testType: TestType) {
+    const hasAccess = await this.accessManagerFactory.forAuthenticatedUser(actor).canAccessUser(test.userId);
```

Comment:
Here I opted for doing access management in the application/service because otherwise I would need to invoke getTest from the repository on presentation if I tried to do it on the api layer